lib/src/Pos/Block/Network/Retrieval.hs

```diff
@@ -283,36 +263,75 @@ dropRecoveryHeader nodeId = do
 -- | Drops recovery header and, if it was successful, queries tips.
 dropRecoveryHeaderAndRepeat
     :: (BlockWorkMode ctx m)
-    => EnqueueMsg m -> NodeId -> m ()
-dropRecoveryHeaderAndRepeat enqueue nodeId = do
+    => Diffusion m -> NodeId -> m ()
+dropRecoveryHeaderAndRepeat diffusion nodeId = do
     kicked <- dropRecoveryHeader nodeId
     when kicked $ attemptRestartRecovery
   where
     attemptRestartRecovery = do
         logDebug "Attempting to restart recovery"
         delay $ sec 2
-        handleAll handleRecoveryTriggerE $ triggerRecovery enqueue
+        handleAll handleRecoveryTriggerE $ triggerRecovery diffusion
         logDebug "Attempting to restart recovery over"
     handleRecoveryTriggerE =
         -- REPORT:ERROR 'reportOrLogE' somewhere in block retrieval.
         reportOrLogE $ "Exception happened while trying to trigger " <>
                        "recovery inside dropRecoveryHeaderAndRepeat: "
 
-----------------------------------------------------------------------------
--- Block request/processing logic
-----------------------------------------------------------------------------
-
 -- Returns only if blocks were successfully downloaded and
 -- processed. Throws exception if something goes wrong.
 getProcessBlocks
     :: forall ctx m.
        (BlockWorkMode ctx m)
-    => EnqueueMsg m
+    => Diffusion m
     -> NodeId
     -> BlockHeader
-    -> HeaderHash
+    -> [HeaderHash]
     -> m ()
-getProcessBlocks enqueue nodeId lcaChild newestHash = do
+getProcessBlocks diffusion nodeId desired checkpoints = do
+    result <- Diffusion.getBlocks diffusion nodeId desired checkpoints
```

Comment:
`getBlocks` from the diffusion layer takes care of the whole deal: get the headers (using checkpoints) and then get the corresponding blocks. I think this is better than having the caller responsible for getting the headers and then the blocks, but of course that's up for debate...

> that protocol (block/header exchange) doesn't have anything to do with diffusion layer. 

Sure it does. Diffusion layer is supposed to abstract this.

Reply:
Now I realize it may be good to split up header and block retrieval rather than doing it all-in-one. It's only because there is some blockchain logic mixed in there at the moment: `matchRequestedHeaders`. Ideally the diffusion layer would just do what is asked of it: get those blocks. But instead it does a sanity check on the headers returned and blows up if either of the ends does not match what was expected.

But I also want to ask the important question: why bother? This seems like a silly check to make. Are we worried about a malicious peer giving us nonsense? They can still do it, just not at the ends. They can give bogus headers in the middle.

Reply:
>`getBlocks` from the diffusion layer takes care of the whole deal: get the headers (using checkpoints) and then get the corresponding blocks. 

It's the first time i hear about this interface. Diffusion layer should be flexible, abstracting data communication methods only IMHO (it's not even an opinion, i thought that's what it was for). `getBlocks` in my opinion ... gets blocks. It doesn't perform any of the recovery logic with header verification. Diffusion layer shouldn't know about the fact headers can be valid or invalid, it also shouldn't consider blockchain state (gstate), except for update system (which define network protocol parameters). It doesn't make sense other way -- what retrievalWorker will become if you put all logic into diffusion layer?  

>I think this is better than having the caller responsible for getting the headers and then the blocks, but of course that's up for debate...

It's better for caller to ask diffusion layer to do all the recovery process, but we're designing both components at the same time, so distinction should help and simplify things only. GState-based verification, header/blocks analysis is what logic does. Diffusion is about "get me this data in most optimal way" or "propagate this data". Please correct me if i misinterpret your original ideas. 

>But instead it does a sanity check on the headers returned and blows up if either of the ends does not match what was expected.

It also does `classifyHeaders`, which is important.

Reply:
> It's the first time i hear about this interface. Diffusion layer should be flexible, abstracting data communication methods only IMHO (it's not even an opinion, i thought that's what it was for). getBlocks in my opinion ... gets blocks. It doesn't perform any of the recovery logic with header verification. Diffusion layer shouldn't know about the fact headers can be valid or invalid, it also shouldn't consider blockchain state (gstate), except for update system (which define network protocol parameters). It doesn't make sense other way -- what retrievalWorker will become if you put all logic into diffusion layer?

I agree. What we're looking at here is in fact a part of the logic layer, and it uses `Diffusion.getBlocks`, which merely gets blocks.

Reply:
Ok, we can leave it as it is. There's https://iohk.myjetbrains.com/youtrack/issue/CSL-1514 and because of that, we want to first query headers and then query blocks if headers are valid. What `getBlocks` from the diffusion layer does now is "unverified" `requestHeaders >> requestBlocks`, which makes CSL-1514 harder to be implemented later. I suggest you to split `getBlocks` in the way it was done before.

Reply:
And after all, again. `getBlocks` shouldn't be placed in `Pos.Diffusion`. 
1. It knows about recovery, and in my opinion recovery is completely logic-related thing.
2. It uses `DialogUnexpected` exception handling, which is logic part from retrieval worker -- you moved it to `Full.Block`, but it's still used by recovery logic.

Reply:
>There is an aspect of block retrieval which is about translocation of the block data. That part must be in the diffusion layer, by definition of the diffusion layer.

It's too ambiguous. What "translocation of the block data" are we talking about? Over the network? From the network to the retrieval queue?

I agree with everything in the second paragraph, though it's not really clear how CSL-1514 will be fixed at the logic level. I still don't get it -- if you agree that recovery is not a part of the diffusion, why not move it out of it? It seems that separating `getBlocks` from `getHeaders` will do for now. It'll be possible to, later, unite them, if we find a way to do it.

>But it's not a problem...

I understand your point of view. But the thing is here not only about perspective, but also about fact that the only way to fix "diffusion doing the classification" is to actually split one diffusion layer in two. The reason for it is current protocol scheme. I also think that downloading and applying blocks directly can do instead, then this verification can be thrown away. Though, it's not clear why this simplification process stops here then. We can respond with blocks when somebody sends checkpoints directly. I suggest to keep the protocol as it is and modify it later in a separate issue, so we don't miss anything (this PR is huge already). 

I'm fine with unsafe `getHeaders >>= getBlocks` for now, but we need to fix it after we merge this PR.